modules/afamqp/afamqp.c

```diff
@@ -654,6 +684,31 @@ afamqp_worker_insert(LogThreadedDestDriver *s, LogMessage *msg)
   return LTR_SUCCESS;
 }
 
+static void
+_handle_heartbeat(void *cookie)
+{
+  AMQPDestDriver *self = (AMQPDestDriver *) cookie;
+
+  amqp_frame_t frame;
+  struct timeval tv = {0, 0};
+  gint status;
+  while (AMQP_STATUS_OK == (status = amqp_simple_wait_frame_noblock(self->conn, &frame, &tv)));
```

Comment:
@gaborznagy
It might be confusing, but amqp_simple_wait_frame_noblock called with {0,0} is non blocking. It becomes blocking if NULL is passed as timeval. From https://alanxz.github.io/rabbitmq-c/docs/0.4.0/amqp_8h.html#a2ad52d60ae5870cc304a39d3a59950ea

"If a 0 timeval is specified, the function behaves as if its non-blocking: it will test to see if a frame can be read from the broker, and return immediately.

If NULL is passed in for tv, the function will behave like amqp_simple_wait_frame() and block until a frame is received from the broker"

Reply:
Yes, what I wanted to point out is that a non-blocking function is called in a while loop.
Of course, it will return immediately and shows if there is a previously received frame or error happened (heartbeat timeout) before calling the function.

Can't we remove the while loop? In a busy connection we would stuck in here.

Reply:
I forgot that in my understanding, most of the time the execution will jump out from the loop, as the wait timeout will be elapsed immediately.

Reply:
@gaborznagy
One call to amqp_simple_wait_frame_noblock can cache multiple frames. In case frames are already available, it does not call read again on the socket, but returns the already extracted first frame from the cache list. This means if the remote sends the heartbeat at a faster frequency than the ivykis timer frequency, that would result in a socket error in the remote after sufficiently long time. This is the condition that I wanted to avoid with the while loop.

In our publish kind of message sending, I do not expect any traffic towards syslog-ng than heartbeat, so in practice, this while loop should exit after two iterations at most, just as you mentioned earlier.

However, if we want to address this error case too, I have the following proposal: I remove the while loop from the callback. Instead of the while-loop I would reschedule the timer with iv_now if the return code is OK. If the return code is TIMEOUT, then I would reschedule the timer with iv_now+freq. That would still look like an infinite loop, but between two iterations, ivykis could schedule another tasks.

In the optimistic case, this means an extra task is scheduled by ivykis in every freq period. However I do not think that would cause any problem. What do you think?

Reply:
Heartbeat implementation in rabbitmq is confusing for me, but I've got the information that it's bi-directional. Is that right?
https://www.rabbitmq.com/heartbeats.html
http://rabbitmq.1065348.n5.nabble.com/AMQP-heartbeat-Expected-behaviors-td20569.html
"both client and server side can send heartbeat frames".

> if the remote sends the heartbeat at a faster frequency than the ivykis timer frequency, that would result in a socket error in the remote after sufficiently long time

So we need to read every frame that's in the cache.
So much for "heartbeat handled internally". See [here](http://alanxz.github.io/rabbitmq-c/docs/0.8.0/amqp_8h.html#a9a186800ece7517245c734d93f18e4a8)
By the way, the [doc](https://www.rabbitmq.com/heartbeats.html) states that *Heartbeat frames are sent about every heartbeat timeout / 2 seconds.*.
We are already reading the frames too late, is this what the while loop resolves?

I would not push the proposed re-work of the iv_timer usage. 
I'll give you an update soon, emptying the frame cache makes this complex.
I just wanted to ask to drop the while loop, and check for STATUS_TIMEOUT and STATUS_OK.
I accept that HEARTBEAT_TIMEOUT can be handled in the error case (I had problem only with the "unexpected" term, not with disconnecting).

Reply:
I've experimented a bit around rabbitmq:
- yes, heartbeats are bidirectional (I've checked with tcpdump)
> By the way, the doc states that Heartbeat frames are sent about every heartbeat timeout / 2 seconds..
> We are already reading the frames too late, is this what the while loop resolves?

Nope, our rabbitmq library negotiates the **interval** and not the timeout.
Therefore, the timer setting is correct.

> So we need to read every frame that's in the cache.
If we accept this statement, then the code does what it had to, there is no open point from my side. Resolving this conversation.